File: mac/AirClip/Core/PeerManager.swift
import Foundation
import Network
import Combine

/// Owns all active peer connections.
///
/// Trust model: a peer is trusted if its airclip_id matches ours and its device_id
/// is still in the local paired-device registry.
/// Public keys are exchanged in the auth handshake and stored on the PeerConnection itself.
@MainActor
final class PeerManager: ObservableObject {
    static let shared = PeerManager()

    @Published var connectedCount: Int = 0
    @Published var connectedDeviceIds: Set<String> = []

    // Authenticated peers keyed by their deviceId.
    private(set) var peers: [String: PeerConnection] = [:]
    // Connections not yet authenticated.
    private var pending: [UUID: PeerConnection] = [:]
    // Prevents duplicate outgoing connections while auth is in flight.
    private var pendingHints: Set<String> = []
    private var heartbeatTimer: Timer?

    private init() {}

    // MARK: - Trust

    func isTrusted(airClipId: String, deviceId: String) -> Bool {
        guard let myAirClipId = AirClipIdentity.shared.airClipId else { return false }
        return airClipId == myAirClipId
    }

    // MARK: - Peer lifecycle

    func addIncoming(_ peer: PeerConnection) {
        pending[peer.id] = peer
    }

    func connectIfNeeded(to endpoint: NWEndpoint, peerHint: String) {
        guard peers[peerHint] == nil else { return }
        guard !pendingHints.contains(peerHint) else { return }
        pendingHints.insert(peerHint)

        let params    = NWParameters.tcp
        let wsOptions = NWProtocolWebSocket.Options()
        wsOptions.autoReplyPing = true
        params.defaultProtocolStack.applicationProtocols.insert(wsOptions, at: 0)

        let conn = NWConnection(to: endpoint, using: params)
        let peer = PeerConnection(connection: conn, isIncoming: false)
        pending[peer.id] = peer
    }

    func registerPeer(_ peer: PeerConnection) {
        guard let deviceId = peer.deviceId else { return }
        pending.removeValue(forKey: peer.id)
        pendingHints.remove(deviceId)

        let device = AirClipIdentity.shared.pairedDevices[deviceId]

        var didRegisterNewPeer = false
        if let existing = peers[deviceId] {
            // Keep incoming (server-side) connection; close duplicate.
            if peer.isIncoming { existing.close(); peers[deviceId] = peer }
            else               { peer.close() }
        } else {
            peers[deviceId] = peer
            didRegisterNewPeer = true
        }
        updateCount()
        if didRegisterNewPeer, let device {
            DeviceNotificationCoordinator.shared.notifyDeviceConnectedOnce(device)
        }
    }

    func disconnectDevice(_ deviceId: String, notifyRemote: Bool = false) {
        pendingHints.remove(deviceId)
        if let peer = peers.removeValue(forKey: deviceId) {
            if notifyRemote {
                peer.sendDeviceRemoved(targetDeviceId: deviceId)
            } else {
                peer.close()
            }
        }
        updateCount()
    }

    func broadcastDeviceRemoval(targetDeviceId: String) {
        for peer in peers.values {
            peer.sendDeviceRemoved(targetDeviceId: targetDeviceId)
        }
    }

    func removePeer(id: UUID) {
        if let peer = pending[id] {
            if let hint = peer.deviceId { pendingHints.remove(hint) }
            pending.removeValue(forKey: id)
        }
        if let key = peers.first(where: { $0.value.id == id })?.key {
            peers.removeValue(forKey: key)
        }
        updateCount()
    }

    func disconnectAll() {
        pending.values.forEach { $0.close() }
        peers.values.forEach   { $0.close() }
        pending.removeAll()
        peers.removeAll()
        pendingHints.removeAll()
        updateCount()
    }

    // MARK: - Count

    private func updateCount() {
        connectedCount     = peers.count
        connectedDeviceIds = Set(peers.keys)
        updateHeartbeatTimer()
        NotificationCenter.default.post(name: .peerCountChanged, object: connectedCount)
    }

    private func updateHeartbeatTimer() {
        if peers.isEmpty {
            heartbeatTimer?.invalidate()
            heartbeatTimer = nil
            return
        }

        guard heartbeatTimer == nil else { return }
        heartbeatTimer = Timer.scheduledTimer(withTimeInterval: 15, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                self?.sendHeartbeats()
            }
        }
    }

    private func sendHeartbeats() {
        for peer in peers.values {
            peer.sendHeartbeat()
        }
    }

    // MARK: - Broadcast

    func broadcast(packet: ClipboardPacket) {
        let myId = AirClipIdentity.shared.deviceId ?? ""
        for (deviceId, peer) in peers where deviceId != myId {
            guard let pk = peer.publicKey else { continue }
            do {
                let (ct, nonce) = try CryptoManager.shared.encrypt(
                    plaintext: packet.encodedJSONString() ?? packet.text,
                    recipientPublicKeyBase64: pk
                )
                peer.sendClip(
                    ciphertext: ct.base64EncodedString(),
                    nonce: nonce.base64EncodedString()
                )
            } catch {}
        }
    }
}

File: android/app/src/main/java/com/airclip/airclip/lan/PeerManager.kt
package com.airclip.airclip.lan

import android.util.Log
import android.os.Handler
import android.os.Looper
import com.google.gson.Gson
import com.airclip.airclip.AirClipIdentity
import com.airclip.airclip.crypto.SealedBox
import com.airclip.airclip.device.DeviceRemovalNotice
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

private const val TAG = "PeerManager"

/**
 * One authenticated LAN peer — either an incoming (server-role) or outgoing (client-role)
 * connection. The [sendFn] and [closeFn] closures abstract over Java-WebSocket vs OkHttp.
 */
data class Peer(
    val deviceId: String,
    val deviceName: String = "",
    val publicKey: String,
    val platform: String = "",
    val wifiNetwork: String? = null,
    val isIncoming: Boolean,
    val connectionId: String = UUID.randomUUID().toString(),
    private val sendFn: (String) -> Unit,
    private val closeFn: () -> Unit
) {
    fun send(json: String) = sendFn(json)
    fun sendRemovalNotice(targetDeviceId: String) =
        send(PeerManager.gson.toJson(DeviceRemovalNotice.payload(targetDeviceId)))
    fun close() = closeFn()
}

/**
 * Central registry of authenticated LAN peers.
 *
 * Trust model: a peer is trusted iff it presents our airclip_id and its device_id
 * is still in the local paired-device registry.
 *
 * Thread-safety: all mutable state lives in [ConcurrentHashMap].
 * [registerPeer] uses @Synchronized to prevent double-registration races.
 */
object PeerManager {

    private val peers        = ConcurrentHashMap<String, Peer>()
    private val pendingHints = ConcurrentHashMap.newKeySet<String>()

    val gson = Gson()

    private val _connectedCount = MutableStateFlow(0)
    val connectedCount: StateFlow<Int> = _connectedCount

    /** Callback invoked when a live inbound clip arrives. Set by SyncEngine. */
    var onInboundClip: ((ciphertext: String, nonce: String, fromDeviceId: String) -> Unit)? = null

    /** Callback invoked when a history_clip message arrives. Set by SyncEngine. */
    var onHistoryClip: ((ciphertext: String, nonce: String, ts: Long, msgId: String, fromDeviceId: String) -> Unit)? = null

    /** Callback invoked when a peer proves it is still reachable. Set by SyncEngine. */
    var onPeerSeen: ((deviceId: String) -> Unit)? = null

    /** Callback invoked when a peer reports foreground or share activity. Set by SyncEngine. */
    var onPeerActivity: ((deviceId: String, timestampMs: Long) -> Unit)? = null

    /** Callback invoked after a peer becomes the active authenticated connection. Set by SyncEngine. */
    var onPeerRegistered: ((Peer) -> Unit)? = null

    /** Callback invoked when another paired device removes this device from the network. */
    var onRemovedFromNetwork: (() -> Unit)? = null

    // ── Trust ─────────────────────────────────────────────────────────────────

    /**
     * Returns true if [theirAirClipId] matches our AirClip ID and [deviceId]
     * has not been removed from the local paired-device registry.
     */
    fun isTrusted(theirAirClipId: String, deviceId: String): Boolean {
        val myAirClipId = AirClipIdentity.airClipId ?: return false
        return theirAirClipId == myAirClipId
    }

    // ── Peer lifecycle ────────────────────────────────────────────────────────

    /**
     * Register an authenticated peer after both sides complete the handshake.
     * If both sides raced to connect simultaneously, keeps the incoming (server-side) peer.
     */
    @Synchronized
    fun registerPeer(peer: Peer) {
        val existing = peers[peer.deviceId]
        if (existing != null) {
            if (existing.isIncoming && !peer.isIncoming) {
                peer.close()
                Log.d(TAG, "Duplicate outgoing discarded for ${peer.deviceId.take(8)}")
                return
            }
            existing.close()
        }
        peers[peer.deviceId] = peer
        pendingHints.remove(peer.deviceId)
        updateConnectedCount()
        onPeerRegistered?.invoke(peer)
        Log.d(TAG, "Peer registered: ${peer.deviceId.take(8)} incoming=${peer.isIncoming} total=${peers.size}")
    }

    fun onPeerDisconnected(deviceId: String, connectionId: String? = null) {
        val current = peers[deviceId]
        if (connectionId == null || current?.connectionId == connectionId) {
            peers.remove(deviceId)
        }
        pendingHints.remove(deviceId)
        updateConnectedCount()
        Log.d(TAG, "Peer disconnected: ${deviceId.take(8)} remaining=${peers.size}")
    }

    fun disconnectDevice(deviceId: String, notifyRemote: Boolean = false) {
        pendingHints.remove(deviceId)
        val peer = peers.remove(deviceId)
        updateConnectedCount()
        if (notifyRemote) {
            peer?.sendRemovalNotice(deviceId)
            if (peer != null) {
                Handler(Looper.getMainLooper()).postDelayed({ peer.close() }, 150)
                return
            }
        }
        peer?.close()
    }

    fun broadcastDeviceRemoval(targetDeviceId: String) {
        peers.values.forEach { peer ->
            runCatching { peer.sendRemovalNotice(targetDeviceId) }
        }
    }

    /** Mark this device_id as connection-in-progress to deduplicate outgoing attempts. */
    fun addPendingHint(deviceId: String): Boolean = pendingHints.add(deviceId)

    fun hasPeer(deviceId: String): Boolean = peers.containsKey(deviceId)
    fun isPending(deviceId: String): Boolean = pendingHints.contains(deviceId)
    fun connectedPeersSnapshot(): List<Peer> = peers.values.toList()

    // ── Sending ───────────────────────────────────────────────────────────────

    /**
     * Encrypt [plaintext] for each authenticated peer's public key and broadcast.
     * Per-peer failures are logged and skipped — other peers still receive.
     */
    fun broadcast(plaintext: ByteArray) {
        peers.values.forEach { peer ->
            try {
                val enc  = SealedBox.encryptForRecipient(plaintext, peer.publicKey)
                val json = gson.toJson(mapOf(
                    "type"       to "clip",
                    "ciphertext" to enc.ciphertextBase64,
                    "nonce"      to enc.nonce,
                ))
                peer.send(json)
            } catch (e: Exception) {
                Log.w(TAG, "broadcast to ${peer.deviceId.take(8)} failed: ${e.message}")
            }
        }
    }

    fun sendHeartbeats() {
        val json = gson.toJson(mapOf(
            "type" to "heartbeat",
            "ts" to System.currentTimeMillis().toString(),
        ))
        peers.values.forEach { peer ->
            runCatching { peer.send(json) }
        }
    }

    fun broadcastAppActivity(timestampMs: Long = System.currentTimeMillis()) {
        val json = gson.toJson(mapOf(
            "type" to "app_activity",
            "ts" to timestampMs.toString(),
        ))
        peers.values.forEach { peer ->
            runCatching { peer.send(json) }
        }
    }

    // ── Receiving ─────────────────────────────────────────────────────────────

    fun dispatchInboundClip(ciphertext: String, nonce: String, fromDeviceId: String) {
        onInboundClip?.invoke(ciphertext, nonce, fromDeviceId)
    }

    fun dispatchHistoryClip(ciphertext: String, nonce: String, ts: Long, msgId: String, fromDeviceId: String) {
        onHistoryClip?.invoke(ciphertext, nonce, ts, msgId, fromDeviceId)
    }

    fun markPeerSeen(deviceId: String) {
        onPeerSeen?.invoke(deviceId)
    }

    fun markPeerActivity(deviceId: String, timestampMs: Long) {
        onPeerActivity?.invoke(deviceId, timestampMs)
    }

    fun dispatchRemovedFromNetwork() {
        onRemovedFromNetwork?.invoke()
    }

    // ── Teardown ──────────────────────────────────────────────────────────────

    fun disconnectAll() {
        peers.values.forEach { runCatching { it.close() } }
        peers.clear()
        pendingHints.clear()
        updateConnectedCount()
    }

    private fun updateConnectedCount() {
        _connectedCount.value = peers.size
    }
}
